{"ast":null,"code":"var attempts = {};\nvar defaultsByInstance = new Map();\nvar networkErrorAttempts = 0;\nexport default function rateLimit(instance, defaults) {\n  var maxRetry = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 5;\n  defaultsByInstance.set(instance, defaults);\n  var instanceDefaults = defaultsByInstance.get(instance);\n  var _instanceDefaults$res = instanceDefaults.responseLogger,\n      responseLogger = _instanceDefaults$res === void 0 ? function () {\n    return undefined;\n  } : _instanceDefaults$res,\n      _instanceDefaults$req = instanceDefaults.requestLogger,\n      requestLogger = _instanceDefaults$req === void 0 ? function () {\n    return undefined;\n  } : _instanceDefaults$req;\n  instance.interceptors.request.use(function (config) {\n    requestLogger(config);\n    return config;\n  }, function (error) {\n    return Promise.reject(error);\n  });\n  instance.interceptors.response.use(function (response) {\n    // we don't need to do anything here\n    responseLogger(response);\n    return response;\n  }, function (error) {\n    var response = error.response,\n        config = error.config; // Do not retry if it is disabled or no request config exists (not an axios error)\n\n    if (!config || !instanceDefaults.retryOnError) {\n      return Promise.reject(error);\n    }\n\n    var retryErrorType = null;\n    var wait = 0; // Errors without response did not recieve anything from the server\n\n    if (!response) {\n      retryErrorType = 'Connection';\n      networkErrorAttempts++;\n\n      if (networkErrorAttempts > maxRetry) {\n        error.attempts = networkErrorAttempts;\n        return Promise.reject(error);\n      }\n\n      wait = Math.pow(Math.SQRT2, networkErrorAttempts);\n      response = {};\n    } else {\n      networkErrorAttempts = 0;\n    }\n\n    if (response.status >= 500 && response.status < 600) {\n      // 5** errors are server related\n      retryErrorType = \"Server \".concat(response.status);\n      var headers = response.headers || {};\n      var requestId = headers['x-contentful-request-id'] || null;\n      attempts[requestId] = attempts[requestId] || 0;\n      attempts[requestId]++; // we reject if there are too much errors of with the same request id\n\n      if (attempts[requestId] > maxRetry || !requestId) {\n        error.attempts = attempts[requestId];\n        return Promise.reject(error);\n      }\n\n      wait = Math.pow(Math.SQRT2, attempts[requestId]);\n    } else if (response.status === 429) {\n      // 429 errors are exceeded rate limit exceptions\n      retryErrorType = 'Rate limit'; // all headers are lowercased by axios https://github.com/mzabriskie/axios/issues/413\n\n      if (response.headers && error.response.headers['x-contentful-ratelimit-reset']) {\n        wait = response.headers['x-contentful-ratelimit-reset'];\n      }\n    }\n\n    var delay = function delay(ms) {\n      return new Promise(function (resolve) {\n        setTimeout(resolve, ms);\n      });\n    };\n\n    if (retryErrorType) {\n      // convert to ms and add jitter\n      wait = Math.floor(wait * 1000 + Math.random() * 200 + 500);\n      instanceDefaults.logHandler('warning', \"\".concat(retryErrorType, \" error occurred. Waiting for \").concat(wait, \" ms before retrying...\"));\n      return delay(wait).then(function () {\n        return instance(config);\n      });\n    }\n\n    return Promise.reject(error);\n  });\n}","map":null,"metadata":{},"sourceType":"module"}